src/Symfony/Component/Messenger/Handler/HandlersLocator.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Handler;
+
+/**
+ * Maps a message to a list of handlers.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class HandlersLocator implements HandlersLocatorInterface
+{
+    private $handlers;
+
+    /**
+     * @param callable[][] $handlers
+     */
+    public function __construct(array $handlers)
+    {
+        $this->handlers = $handlers;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getHandlers(string $name): iterable
+    {
+        $seen = array();
+
+        foreach (self::listTypes($name) as $type) {
+            foreach ($this->handlers[$type] ?? array() as $handler) {
+                if (!\in_array($handler, $seen, true)) {
+                    yield $seen[] = $handler;
+                }
+            }
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function listTypes(string $class): array
+    {
+        if (!class_exists($class, false)) {
+            return array($class => $class, '*' => '*');
+        }
+
+        return array($class => $class)
```

Comment:
The array key does not seem to be used or useful. So you could put an `array_values` around the whole array union. This would clarify the code to me as it makes clear the array key is only relevant for the array union operation.

Reply:
`return array_values(array($class => $class) + ...`

Reply:
An array_values() call would add memory and CPU overhead for no reason. Yes, we don't use the keys. That's not an issue to me - arrays always have keys anyway. And actually having the key be the values makes it impossible to return twice the same value, which is part of the contract of this function. I'd keep as is.